app/adapters/bagit/query_service.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+module Bagit
+  class QueryService
+    attr_reader :adapter
+    def initialize(adapter:)
+      @adapter = adapter
+    end
+
+    def find_by(id:)
+      raise ArgumentError, 'id must be a Valkyrie::ID' unless id.is_a? Valkyrie::ID
+      loader = Bagit::BagLoader.new(adapter: adapter, id: id)
+      raise Valkyrie::Persistence::ObjectNotFoundError unless loader.exist?
+      loader.load!
+    end
+
+    def find_all
+      adapter.bag_paths.lazy.map do |bag_path|
+        find_by(id: Valkyrie::ID.new(Pathname.new(bag_path).basename))
+      end
+    end
+
+    def find_members(resource:, model: nil)
+      find_references_by(resource: resource, property: :member_ids).select do |member|
+        model.nil? || member.is_a?(model)
+      end
+    end
+
+    def find_parents(resource:)
+      find_inverse_references_by(resource: resource, property: :member_ids)
+    end
+
+    def find_references_by(resource:, property:)
+      ids = (resource.try(property) || []).select { |id| id.is_a?(Valkyrie::ID) }
+      ids.lazy.map do |id|
+        find_by(id: id)
+      end
+    end
+
+    def find_all_of_model(model:)
+      Valkyrie.logger.warn("Bagit Query Service has been asked to find all resources of a specific type. This will require iterating over the metadata of every bag - AVOID.")
```

Comment:
I like these warnings — I wonder if we should add them to the Fedora query service methods that use `find_all` too?

Reply:
Yeah, probably not a bad idea.